### PR TITLE
III-4685 prevent identical tariffs for imports

### PR DESCRIPTION
**Pull request by JonasVHG**
### Added

- Add a check on duplicate `Tariff` names for imports

---
Ticket: https://jira.uitdatabank.be/browse/III-4685

### CODE REVIEW

File: tests/Model/ValueObject/Price/TariffsTest.php
Comment (LucWollants) [inline tests/Model/ValueObject/Price/TariffsTest.php:88]: I think this name is a bit misleading, because it does not find names in another language.

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (bertramakers) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:20]: Nit: Since this class only gets executed after `JsonSchemaValidatingRequestBodyParser`, the `priceInfo` property should always be an array if present. So this if could be reduced to just:
```
if (!isset($data->priceInfo)) {
    return $request; 
}
```

(In some other RequestBodyParsers we also check the type and if other subproperties are set and the correct type etc because those are called _before_ `JsonSchemaValidatingRequestBodyParser` to transform old/legacy JSON request bodies to the ones expected by the JSON schema.)

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (bertramakers) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:34]: Would be cool if we could have a more specific jsonPointer here that points to the tariff name that is a duplicate. (E.g. `/priceInfo/2/name` if the 3rd tariff's name is a duplicate of an earlier one)

Also nice would be to include multiple schemaErrors if there are multiple duplicates, so the integrator doesn't need to fix them one by one.

File: src/Http/Event/ImportEventRequestHandler.php
Comment (bertramakers) [inline src/Http/Event/ImportEventRequestHandler.php:115]: @LucWollants @JonasVHG maybe we should start looking into combining some of these into a class like `OfferValidatingRequestBodyParser` so we don't have to maintain 2 lists of property validators (here + `ImportPlaceRequestHandler`). Can be in another PR later though.

File: src/Model/ValueObject/Price/Tariffs.php
Comment (bertramakers) [inline src/Model/ValueObject/Price/Tariffs.php:17]:
```suggestion
    public function hasDuplicateNames(): bool
```
Small nit but this might be clearer, as a duplicate can also be interpreted as same name + same price + same currency.

**Review by bertramakers — COMMENTED** [review]
Thanks @JonasVHG, overall approach looks great! 👌 Just some small suggestions.

File: src/Http/Event/ImportEventRequestHandler.php
Comment (LucWollants) [reply src/Http/Event/ImportEventRequestHandler.php:115]: Yes, that can certainly help!

File: tests/Http/Offer/PriceInfoValidatingRequestBodyParserTest.php
Comment (LucWollants) [inline tests/Http/Offer/PriceInfoValidatingRequestBodyParserTest.php:46]: I don't see a throw here?

File: tests/Http/Event/ImportEventRequestHandlerTest.php
Comment (LucWollants) [inline tests/Http/Event/ImportEventRequestHandlerTest.php:4416]: Maybe the base category should have another index to make sure that case is also handled?

File: src/Http/Place/ImportPlaceRequestHandler.php
Comment (LucWollants) [inline src/Http/Place/ImportPlaceRequestHandler.php:118]: This is not tested I think?

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:41]: Maybe change this to `foreach ($priceInfos as $index => $priceInfo)` for readability?

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:42]: I think the correct naming is `foreach ($priceInfo['name'] as $language => $name)`

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [reply src/Http/Offer/PriceInfoValidatingRequestBodyParser.php]: Should be fixed now

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [reply src/Http/Offer/PriceInfoValidatingRequestBodyParser.php]: @JonasVHG this is maybe something that can still be done inside this pull request?

File: src/Model/ValueObject/Price/Tariffs.php
Comment (LucWollants) [reply src/Model/ValueObject/Price/Tariffs.php]: Method is no longer used.

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [reply src/Http/Offer/PriceInfoValidatingRequestBodyParser.php]: This is changed.

File: src/Http/Event/ImportEventRequestHandler.php
Comment (LucWollants) [reply src/Http/Event/ImportEventRequestHandler.php:115]: @JonasVHG this is maybe something that can still be done inside this pull request?

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:42]: I think you forgot to change `$priceLang` to `$name`?

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (LucWollants) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:39]: After changing the variable names, I wonder this is not a `$nameMatrix` instead of a `$priceMatrix`?

File: src/Http/Event/ImportEventRequestHandler.php
Comment (JonasVHG) [reply src/Http/Event/ImportEventRequestHandler.php:115]: @LucWollants , I've tried it but but I'm afraid that it causes too much code changes so I assume that it would be better in a different PR.

File: src/Http/Offer/PriceInfoValidatingRequestBodyParser.php
Comment (bertramakers) [inline src/Http/Offer/PriceInfoValidatingRequestBodyParser.php:47]:
```suggestion
                        'Tariff name "' . $priceLang . '" must be unique.'
```
Just a small nit but maybe clearer since SHOULD is often seen as a recommendation, and MUST as a requirement in technical docs.

File: tests/Http/Offer/PriceInfoValidatingRequestBodyParserTest.php
Comment (bertramakers) [reply tests/Http/Offer/PriceInfoValidatingRequestBodyParserTest.php:61]: It does `$this->assertCallableThrowsApiProblem()` on the last line of the test. But maybe that was changed after your comment, I haven't reviewed the whole history of commits just the overall diff.

**Review by bertramakers — COMMENTED** [review]
I leave it up to @LucWollants to approve after he has checked that all his comments have been resolved with the latest changes, but for me this looks great! Just added a small nitpick but nothing super important.

File: src/Http/Event/ImportEventRequestHandler.php
Comment (LucWollants) [reply src/Http/Event/ImportEventRequestHandler.php:127]: Done inside https://github.com/cultuurnet/udb3-silex/pull/1014

**Review by LucWollants — APPROVED** [review]

**Pull request merged**
